generators/server/templates/src/main/java/package/security/oauth2/AuthorizationHeaderUtil.java.ejs

```diff
@@ -18,46 +18,162 @@
 import org.springframework.stereotype.Component;
+import org.springframework.util.LinkedMultiValueMap;
+import org.springframework.util.MultiValueMap;
+import org.springframework.web.client.RestTemplate;
 
-import java.util.Optional;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
 
 @Component
 public class AuthorizationHeaderUtil {
 
     private final OAuth2AuthorizedClientService clientService;
+    private final RestTemplateBuilder restTemplateBuilder;
+    private final JwtDecoder jwtDecoder;
+    private final Logger log = LoggerFactory.getLogger(AuthorizationHeaderUtil.class);
 
-    public AuthorizationHeaderUtil(OAuth2AuthorizedClientService clientService) {
+    public AuthorizationHeaderUtil(OAuth2AuthorizedClientService clientService, RestTemplateBuilder restTemplateBuilder, JwtDecoder jwtDecoder) {
         this.clientService = clientService;
+        this.restTemplateBuilder = restTemplateBuilder;
+        this.jwtDecoder = jwtDecoder;
     }
 
     public Optional<String> getAuthorizationHeader() {
         Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
-        String accessToken = null;
-        AbstractAuthenticationToken authToken = (AbstractAuthenticationToken) authentication;
-        if (authToken instanceof OAuth2AuthenticationToken) {
-            OAuth2AuthenticationToken oauthToken = (OAuth2AuthenticationToken) authToken;
+        if (authentication instanceof OAuth2AuthenticationToken) {
+            OAuth2AuthenticationToken oauthToken = (OAuth2AuthenticationToken) authentication;
+            String name =  oauthToken.getName();
+            String registrationId = oauthToken.getAuthorizedClientRegistrationId();
             OAuth2AuthorizedClient client = clientService.loadAuthorizedClient(
-                oauthToken.getAuthorizedClientRegistrationId(),
-                oauthToken.getName());
-            accessToken = client.getAccessToken().getTokenValue();
-        } else if (authToken instanceof JwtAuthenticationToken) {
-            accessToken = ((JwtAuthenticationToken) authToken).getToken().getTokenValue();
-        }
+            registrationId,
+            name);
+
+            if (null == client) {
+                throw new OAuth2AuthorizationException(new OAuth2Error("access_denied", "The token is expired", null));
+            }
+            OAuth2AccessToken accessToken = client.getAccessToken();
+
+            if (accessToken != null) {
+                String tokenType = accessToken.getTokenType().getValue();
+                String accessTokenValue = accessToken.getTokenValue();
+                if (isExpired(accessToken)) {
+                    log.info("AccessToken expired, refreshing automatically");
+                    accessTokenValue = refreshToken(client, oauthToken);
+                    if (null == accessTokenValue) {
+                        SecurityContextHolder.getContext().setAuthentication(null);
```

Comment:
I am not very sure on this statement. Can you verify the behavior if you explicitly set to null, but, still authenticated with provider (session cookie)?

Reply:
Looks like it is working, at least on all my use case. Also, I really don't know what could be an alternative